### DISCUSSION TITLE
[Question] Undelete an annotation

### DISCUSSION THREAD

**Discussion by PowerSnail** [pymupdf/PyMuPDF#900] [Q&A] [answered]
The doc of delete_annot says:

> Please remember that physical removal requires saving to a new file with garbage > 0.

Does that mean that there is an easy way to un-delete the annotation, while the file hasn't been saved?

Of course I can re-insert a new annotation, with the same attributes, but the problem is that the order of annotation will not be the same.

Thanks!

**Comment by JorjMcKie** [comment]
> Does that mean that there is an easy way to un-delete the annotation, while the file hasn't been saved?

1. If you `doc.close()` and then re-open (for example by `doc = fitz.open(doc.name)`), the document is in the **_state before all changes_** since the last save. You cannot undo some changes, but accept others.
2. If you do a `doc.saveIncr()´ or `doc.save(doc.name, incremental=True,...)`, the document will **_contain_** your changes. But physically the previous document state continues to exist. There is no way to go back to an earlier version via PyMuPDF, but I quite often use a capable editor to a remove recent incremental updates. Fairly easy thing to do. And of course, with incremental updates, the file size will always grow.
3. If you do not use incremental updates, deleted stuff does not go away physically if you use `garbage=0`. The file size will also grow, but earlier viersions cannot be recovered. To recover space and to remove potentially sensitive data also physically, use a garbage collection value > 0. Please look up the documentation to learn about the differences between values 1 to 4.

**Reply by PowerSnail** [reply]
I see. Thanks for the answer.

A follow-up question: is it possible to change the order of annotations?

**Reply by JorjMcKie** [reply]
> is it possible to change the order of annotations?

You mean as they appear in `page.annots()`?
No, this sequence is technically taken from a PDF array in the page's object definition found under the PDF key `/Annots`.
In this array, the xrefs of not only annotations are stored, but also those for links and fields on the page.
It generally looks like this:
```postscript
  /Annots [ 5 0 R 9 0 R 12 0 R 13 0 R 17 0 R 21 0 R 25 0 R 29 0 R
      32 0 R 35 0 R 38 0 R 41 0 R 44 0 R 47 0 R 48 0 R 52 0 R
      57 0 R ]
```
Xrefs of new annotations (or links, fields) are **_appended_** to this array as a string of the form ``xref 0 R``.
Presumably you would like to see a sequence in some "natural" reading order: like left-to-right, top-to-bottom.
You can however make your annot list and sort that as you whish:

```python
annots = [a for a in page.annots()]
annots.sort(key=lambda a: (a.rect.tl.y, a.rect.tl.x))
```
The annot list is sorted by the annotations' top-left corner vertically, then horizontally giving the previously mentioned reading order.
Ask an Arabian friend if he / she finds this "natural" 😉.

**Reply by PowerSnail** [reply]
I wanted to change the order, because if two annotations are overlapping with each other, and I delete the first one, then undo the delete by re-inserting the annotation, the bottom annotation is on top now.

I suppose that the only solution for me is to make my own pseudo-delete where I just hide some annotations instead of calling into pymupdf's delete.

Thanks, though. Your explanation clarifies it for me.

**Comment by JorjMcKie** [comment, marked as answer]
Ah, now I see what you want.
There is a low-level code approach available: you can directly change the page object definition.
Edit the text string returned by ``doc.xref_object(page.xref)`` so that the two annotation xrefs change their sequence and replace the page definition with the resulting string.

```python
page_obj = doc.xref_object(page.xref)
# manipulate string page_obj, then:
doc.update_object(page.xref, page_obj)
# do not forget:
page = doc.reload_page(page)
```

The newest PyMuPDF version also allows a more direct change of the /Annots key content – see `doc.xref_get_key()` and `doc.xref_set_key()`.
Example session:
```python
>>> page = doc[0]
>>> annots = doc.xref_get_key(page.xref, "Annots") # get string of pages annotations
>>> annots
('array', '[5 0 R 9 0 R 12 0 R 13 0 R 17 0 R 21 0 R 25 0 R 29 0 R 32 0 R 35 0 R 38 0 R 41 0 R 44 0 R 47 0 R 48 0 R 52 0 R 57 0 R]')
>>> # assume we want to exchange places of xref 17 and 35
>>> annots_list = annots[1][1:-1].replace(" 0 R","")
>>> annots_list
'5 9 12 13 17 21 25 29 32 35 38 41 44 47 48 52 57'
>>> annots_list = annots_list.split()
>>> # exchange positions of 17 and 35:
>>> i1 = annots_list.index("17")
>>> i2 = annots_list.index("35")
>>> annots_list[i1], annots_list[i2] = annots_list[i2], annots_list[i1]
>>> annots_text = "[" +  " 0 R ".join(annots_list) + " 0 R]
>>> annots_text  # confirm it worked
'[5 0 R 9 0 R 12 0 R 13 0 R 35 0 R 21 0 R 25 0 R 29 0 R 32 0 R 17 0 R 38 0 R 41 0 R 44 0 R 47 0 R 48 0 R 52 0 R 57 0 R]'
>>> doc.xref_set_key(page.xref, "Annots", annots_text)  # store back to the page
>>> page = doc.reload_page(page)  # reload the page
>>> print(doc.xref_object(page.xref))  # and look at changed annot sequence
<<
  /Type /Page
  /MediaBox [ 0 0 595 842 ]
  /Rotate 0
  /Resources 3 0 R
  /Parent 2 0 R
  /Annots [ 5 0 R 9 0 R 12 0 R 13 0 R 35 0 R 21 0 R 25 0 R 29 0 R
      32 0 R 17 0 R 38 0 R 41 0 R 44 0 R 47 0 R 48 0 R 52 0 R
      57 0 R ]
  /Contents [ 8 0 R 11 0 R 15 0 R 16 0 R 19 0 R 20 0 R 23 0 R
      24 0 R 27 0 R 28 0 R 31 0 R 34 0 R 37 0 R 40 0 R 43 0 R
      46 0 R 51 0 R 55 0 R 56 0 R 59 0 R ]
>>
>>> 
```

**Reply by PowerSnail** [reply]
I see. Thanks for the reply. The PDF format is a lot more complicated than I thought. 

To clarify, "0 R" is just a delimiter right?

**Reply by JorjMcKie** [reply]
> To clarify, "0 R" is just a delimiter right?

Right. The "0" is actually a version number of the object. But it only is != 0 in delta data when the PDF is incrementally updated.

**Reply by PowerSnail** [reply]
Thanks! That's been very helpful.
